python/ray/serve/_private/long_poll.py

```diff
@@ -398,10 +408,18 @@ def notify_changed(
         object_key: KeyType,
         updated_object: Any,
     ):
-        self.snapshot_ids[object_key] += 1
+        try:
+            self.snapshot_ids[object_key] += 1
+        except KeyError:
+            # Initial snapshot id must be >= 0, so that the long poll client
+            # can send a negative initial snapshot id to get a fast update.
+            # They should also be randomized to try to avoid situations where,
+            # if the controller restarts and a client has a now-invalid snapshot id
+            # that happens to match what the controller restarts with,
+            # the client wouldn't receive an update.
+            self.snapshot_ids[object_key] = random.randint(0, 1_000_000)
```

Comment:
Now that this isn't a `defaultdict` anymore, could you double check whether any other places we access `self.snapshot_ids` needs this `KeyError` check?

Reply:
I see four uses:

The first two are in https://github.com/ray-project/ray/blob/fd37d93270bace200e1d1ac274fb38314f4439f1/python/ray/serve/_private/long_poll.py#L411-L418 , which are this block that's protected by the `KeyError` check.

The next is https://github.com/ray-project/ray/blob/fd37d93270bace200e1d1ac274fb38314f4439f1/python/ray/serve/_private/long_poll.py#L256-L264 , which is also protected by a `KeyError` check.

The last is https://github.com/ray-project/ray/blob/fd37d93270bace200e1d1ac274fb38314f4439f1/python/ray/serve/_private/long_poll.py#L308-L317 . That block runs when a key changes, which is triggered by https://github.com/ray-project/ray/blob/fd37d93270bace200e1d1ac274fb38314f4439f1/python/ray/serve/_private/long_poll.py#L422-L423, which is below that first block which ensure the snapshot id is present in the mapping, so I think we're good on that one too.

Reply:
Great, thanks for being thorough here!

Reply:
My pleasure!